File: Planarian/Planarian/Modules/Caves/Models/CaveChangeRequestVm.cs
using Planarian.Model.Database.Entities.RidgeWalker;
using Planarian.Model.Database.Revisions;

namespace Planarian.Modules.Caves.Models;

public sealed record CaveChangeRequestSummaryVm(
    string Id,
    string CaveId,
    string CaveName,
    bool CaveExists,
    CaveChangeRequestStatus Status,
    string SubmitterUserId,
    string? SubmitterName,
    DateTime SubmittedOn,
    DateTime? UpdatedOn,
    DateTime? ReviewedOn,
    string? ReviewerUserId,
    string? ReviewerName,
    string? ReviewerNotes,
    string OriginalBaseRevisionId,
    string ProposalBaseRevisionId,
    string? CurrentRevisionId,
    string CurrentProposalVersionId,
    bool IsStale,
    string? ApprovedRevisionId,
    bool CanEdit,
    bool CanReview);

public sealed record CaveChangeRequestDetailVm(
    CaveChangeRequestSummaryVm Request,
    CavePublishedSnapshotV1 Base,
    CavePublishedSnapshotV1 Proposed,
    CavePublishedSnapshotV1 Current,
    CaveRevisionDiffVm Diff,
    CaveRevisionDiffVm? PublishedSinceBase,
    IReadOnlyList<CaveProposalVersionVm> Versions,
    CountyNumberIntent CountyNumberIntent,
    int? RequestedCountyNumber,
    IReadOnlyList<CaveFileSnapshotV1> ActiveStagedFiles);

public sealed record CaveChangePreviewVm(CavePublishedSnapshotV1 Base,
    CavePublishedSnapshotV1 Proposed, CaveRevisionDiffVm Diff,
    CountyNumberIntent CountyNumberIntent, int? RequestedCountyNumber);

public sealed record CreateCaveChangeRequestVm(AddCaveVm Cave, string ExpectedBaseRevisionId);

public sealed record CaveProposalAuthoringContextVm(CaveVm Cave, string ExpectedBaseRevisionId);

public sealed record ReviseCaveChangeRequestVm(AddCaveVm Cave, string ExpectedBaseRevisionId,
    string ExpectedProposalVersionId, bool AgainstCurrent);

public enum CaveProposalAuthoringConflictKind
{
    PublishedCaveChanged,
    ActiveProposalVersionChanged
}

public sealed record CaveProposalAuthoringConflictVm(CaveProposalAuthoringConflictKind ConflictKind,
    string ExpectedId, string? ActualId);

public sealed record CaveProposalVersionVm(string Id, string? PreviousProposalVersionId,
    string BaseRevisionId, string? CreatedByUserId, string? CreatedByName, DateTime CreatedOn, bool IsCurrent);

public enum CaveProposalCountyNumberMode
{
    PreserveExisting,
    Manual,
    FirstAvailable,
    AutomaticNext
}

public sealed record CaveProposalCountyNumberStateVm(CaveProposalCountyNumberMode Mode, int? Number = null);

public static class CaveProposalCountyNumberPresentation
{
    public static CaveProposalCountyNumberStateVm Resolve(CaveProposalSnapshotV1 proposal,
        CavePublishedSnapshotV1 baseSnapshot)
    {
        if (proposal.CountyNumberIntent == CountyNumberIntent.Manual)
            return new CaveProposalCountyNumberStateVm(CaveProposalCountyNumberMode.Manual,
                proposal.RequestedCountyNumber);
        if (proposal.CountyId == baseSnapshot.County.Id)
            return new CaveProposalCountyNumberStateVm(CaveProposalCountyNumberMode.PreserveExisting,
                baseSnapshot.CountyNumber);
        return new CaveProposalCountyNumberStateVm(
            proposal.CountyNumberIntent == CountyNumberIntent.FirstAvailable
                ? CaveProposalCountyNumberMode.FirstAvailable
                : CaveProposalCountyNumberMode.AutomaticNext);
    }
}

public sealed record CaveProposalCountyNumberChangeVm(
    CaveProposalCountyNumberStateVm Previous,
    CaveProposalCountyNumberStateVm Current);

public sealed record CaveProposalVersionDetailVm(CavePublishedSnapshotV1 Base,
    CavePublishedSnapshotV1 Proposed, CaveRevisionDiffVm DiffFromBase,
    CavePublishedSnapshotV1? PreviousProposed, CaveRevisionDiffVm? DiffFromPreviousVersion,
    CaveProposalCountyNumberChangeVm? CountyNumberChange,
    bool BaseRevisionChanged, string? PreviousBaseRevisionId, string BaseRevisionId,
    CountyNumberIntent CountyNumberIntent, int? RequestedCountyNumber,
    IReadOnlyList<string> UnavailableStagedFileIds);

public sealed record CaveChangeRequestDecisionRequestVm(string? Notes, string ExpectedProposalVersionId);

public enum CaveChangeRequestDecisionResult
{
    Approved,
    Rejected,
    Conflict
}

public sealed record CaveChangeRequestDecisionVm(CaveChangeRequestDecisionResult Result,
    string RequestId, string? PublishedRevisionId, string? CurrentRevisionId);

File: Planarian.Web/src/Modules/Caves/Models/CaveChangeRequestVm.ts
import { GeoJsonUploadVm } from "./GeoJsonUploadVm";
import { CaveRevisionDiffVm, CaveSnapshotVm } from "./CaveRevisionVm";

export type CountyNumberIntent = "AutomaticNext" | "FirstAvailable" | "Manual";

export type CaveChangeRequestStatus = "Pending" | "Approved" | "Rejected";

export interface CaveChangeRequestSummaryVm {
  id: string;
  caveId: string;
  caveName: string;
  caveExists: boolean;
  status: CaveChangeRequestStatus;
  submitterUserId: string;
  submitterName?: string;
  submittedOn: string;
  updatedOn?: string;
  reviewedOn?: string;
  reviewerUserId?: string;
  reviewerName?: string;
  reviewerNotes?: string;
  originalBaseRevisionId: string;
  proposalBaseRevisionId: string;
  currentRevisionId?: string;
  currentProposalVersionId: string;
  isStale: boolean;
  approvedRevisionId?: string;
  canEdit: boolean;
  canReview: boolean;
}

export interface CaveProposalVersionVm {
  id: string;
  previousProposalVersionId?: string;
  baseRevisionId: string;
  createdByUserId?: string;
  createdByName?: string;
  createdOn: string;
  isCurrent: boolean;
}

export interface CaveProposalAuthoringContextVm {
  cave: import("./CaveVm").CaveVm;
  expectedBaseRevisionId: string;
  linePlots: GeoJsonUploadVm[];
}

export interface CaveEditAuthoringContextVm {
  cave: import("./CaveVm").CaveVm;
  linePlots: GeoJsonUploadVm[];
}

export interface CaveProposalVersionDetailVm {
  base: CaveSnapshotVm;
  proposed: CaveSnapshotVm;
  diffFromBase: CaveRevisionDiffVm;
  previousProposed?: CaveSnapshotVm;
  diffFromPreviousVersion?: CaveRevisionDiffVm;
  countyNumberChange?: CaveProposalCountyNumberChangeVm;
  baseRevisionChanged: boolean;
  previousBaseRevisionId?: string;
  baseRevisionId: string;
  countyNumberIntent: CountyNumberIntent;
  requestedCountyNumber?: number;
  linePlots: GeoJsonUploadVm[];
  unavailableStagedFileIds: string[];
}

export interface CaveProposalCountyNumberChangeVm {
  previous: CaveProposalCountyNumberStateVm;
  current: CaveProposalCountyNumberStateVm;
}

export interface CaveProposalCountyNumberStateVm {
  mode: "PreserveExisting" | "Manual" | "FirstAvailable" | "AutomaticNext";
  number?: number;
}

export interface CaveChangeRequestDetailVm {
  request: CaveChangeRequestSummaryVm;
  base: CaveSnapshotVm;
  proposed: CaveSnapshotVm;
  current: CaveSnapshotVm;
  diff: CaveRevisionDiffVm;
  publishedSinceBase?: CaveRevisionDiffVm;
  versions: CaveProposalVersionVm[];
  countyNumberIntent: CountyNumberIntent;
  requestedCountyNumber?: number;
  activeStagedFiles: CaveSnapshotVm["files"];
  unavailableStagedFileIds: string[];
}

export interface CaveChangeRequestDecisionVm {
  result: "Approved" | "Rejected" | "Conflict";
  requestId: string;
  publishedRevisionId?: string;
  currentRevisionId?: string;
}

export interface CaveChangePreviewVm {
  base: CaveSnapshotVm;
  proposed: CaveSnapshotVm;
  diff: CaveRevisionDiffVm;
  countyNumberIntent: CountyNumberIntent;
  requestedCountyNumber?: number;
}

export interface CaveProposalAuthoringConflictVm {
  conflictKind: "PublishedCaveChanged" | "ActiveProposalVersionChanged";
  expectedId: string;
  actualId?: string;
}
